### ISSUE TITLE
Support Branch

### ISSUE THREAD

**Issue by macpharlan** [petervanderdoes/gitflow-avh#344]
I am new to gitflow and can find very little about the support branch feature, but what I do know makes me think I could really use it. We have a need to have a support branch for prior releases in order to hotfix special customers, this branch will come off master and never return, this I believe is the idea of support branches.

Questions:
Is it supported now and officially released?

How does it work, this is what I am hoping the workflow looks like:
1) support branch off of master (support/rel1.0.0_support or something)
2) hotfixes can be created off of this support branch
3) when a hotfix is closed/released it should merge to support branch AND developer AND master

Any help with this would be grand!

Thanks,

**Comment by hoall** [comment]
Hey,

knowing the answer is pretty late, but can you please check out the following command?

`git flow --version`

At least since version 1.9.1 (AVH Edition) there're support branches (Ubuntu default). Try

`git flow help`

I know that Suse is still coming with the original version from nvie, which is pretty outdated by now.

**Comment by macpharlan** [comment]
ok that is good to know, I thought it was still in beta.

My --version or help commands don't give the version, I have whatever is
installed with Git 2.13.3.windows.1

C:\workspace\GIT\GIT_Repos>git flow --version
usage: git flow <subcommand>

Available subcommands are:
   init      Initialize a new git repo with support for the branching model.
   feature   Manage your feature branches.
   bugfix    Manage your bugfix branches.
   release   Manage your release branches.
   hotfix    Manage your hotfix branches.
   support   Manage your support branches.
   version   Shows version information.
   config    Manage your git-flow configuration.
   log       Show log deviating from base branch.

Try 'git flow <subcommand> help' for details.

C:\workspace\GIT\GIT_Repos>git flow help
usage: git flow <subcommand>

Available subcommands are:
   init      Initialize a new git repo with support for the branching model.
   feature   Manage your feature branches.
   bugfix    Manage your bugfix branches.
   release   Manage your release branches.
   hotfix    Manage your hotfix branches.
   support   Manage your support branches.
   version   Shows version information.
   config    Manage your git-flow configuration.
   log       Show log deviating from base branch.

Try 'git flow <subcommand> help' for details.

**Comment by jkankiewicz** [comment]
Unfortunately, support branches are still second-class citizens because you can neither publish (#356) nor track (#368) one without resorting to bare `git` command usage. :frowning_face:

**Comment by jkankiewicz** [comment]
> when a hotfix is closed/released it should merge to support branch AND developer AND master

It **should** be automatically merged into the support branch upon which it was based but I can think of a couple of reasons why it **shouldn't** be automatically merged into the develop or master branches:
* Any back-ported fixes that it contains can cause conflicts during a merge.
* It will contain an incremented version number that *precedes* the version number of the latest release because a hotfix branch serves as both a bugfix and a release branch rolled into one.

In general, a hotfix to a support branch is best carried forward by manually cherry-picking commits into a conventional bugfix branch. The version number increment in the hotfix branch should be in a separate commit so that it's easy to exclude it.

**Comment by AlexGoris-Cashforce** [comment]
Hi everyone,
Sorry for using this old issue, but since it was still open I decided it was better than opening a new issue.
I think that @macpharlan asked for in this issue, highlights a bigger issue, namely that the gitflow support branches are not documented at all. All I could find on this repo, is documentation around the `git flow <command> support` commands, but not what they are and how they should be used.

As I am implementing gitflow for our company, and we have a need for supporting multiple versions in parallel, I would be willing to fork this repo and make a PR which documents the actual support branches themselves. Only if we have documented how they're supposed to work, we should discuss how the tooling should behave in various situations.

Would the maintainers of this repo (@luiscoms and @petervanderdoes ) be open for such a PR?

Also, before starting, I would very much appreciate any links to articles, or personal insights about the following points regarding support branches:
* How do we get changes into support branches? Via feature-ish branches? Or are these always hotfixes by design? Maybe dedicated `bugfix/*` branches? Or do we commit directly on them ,and tag specific commits with release versions?
* Pertaining to the above, do we tag releases on these branches directly? I don't see a a way to get changes back into master 
* What happens with changes we do in support branches, in some cases they might need to be merged back into develop (or into a release branch, if we're in the middle of one), but this likely not always possible. 
* It feels like the creation of support branches (provided we make them on the minor version, so `support/1.1.x`) would remove the need for a hotfix branch completely, as you're effectively using master only for minor releases

**Comment by Shea690901** [comment]
As I see it, support branches are for LTS versions.
You might (hot|bug)fix the current release within the master branch, but you also have some LTS versions which are (hot|bug)fixed within their specific support branches.
If you where to fix those LTS versions within the master branch, you would have to rebase master and develop (and every branch depending on either one) on the fix for the LTS version! Something which isn't advisable and in fact might break other things in later releases, regardless if they are LTS or current (not counting the problems which arise with rebaseing anyway, especially when the base commit you're rebaseing on lies hundreds of commits back (develop-branch and offsprings))...
As such support branches are never back merged into master and just die when the end of life time for the corresponding LTS version has been reached. When this time comes you might delete the branches (loosing all work done for those releases) or (IMHO better approach) just let them sit indefinitely (perhaps you even find some idea to fix a bug in some later releases)...

PS.:
Forgot to answer your second point:
If any of those fixes is needed in later versions, you would cherry pick those fixes into the later versions....

**Comment by askus0214** [comment]
halo guys, can you help me?

**Comment by fatso83** [comment]
I added a description of what the support command is for in the [FAQ](https://github.com/petervanderdoes/gitflow-avh/wiki/FAQ). Maybe this could be closed.

**Comment by srevinsaju** [comment]
Apprently, the `git flow support` seems pretty confusing. Here is my scenario:

* My main branch is on v1.x.y
* I created a git flow support branch `REL0_X` for v0.x.x support (last release v0.6.0)
* Now I create a feature branch `feature_branch_1` from `REL0_X`
* I create another feature branch `feature_branch_2` from `REL0_X`
* Now I merge one of them `feature_branch_1` to `REL0_X`
* Now I start the release `git flow release start` creating `v0.7.0`
* My QA team goes through the v0.7.0 and signs it ready
* Before finishing the release, I merge my `feature_branch_2` to back to `REL0_X`
* Now I finish my release `git flow release finish` and v0.7.0
* Taking a look at the git log, I find that the code from `feature_branch_2` is also in the tag when it got merged! But the QA team never tested that as part of v0.7.0 testing 

Am I missing something from this? Shouldn't support branches also have a main+develop pair?

**Comment by jkankiewicz** [comment]
> Am I missing something from this? Shouldn't support branches also have a main+develop pair?

@srevinsaju, a support branch should *not* have a main+develop pair; a support branch is for applying hotfixes to a previous release. Only hotfix branches should be started from a support branch. Any changes to the support branch that should be carried forward to the main branch should be cherry-picked into a bugfix or feature branch.

In your scenario, `REL0_X` should be the main branch in a separate repository.

**Comment by srevinsaju** [comment]
Right, that makes sense, thanks for confirming.
